src/algebra/module/conjugate.lean

```diff
@@ -0,0 +1,121 @@
+/-
+Copyright (c) 2020 Frédéric Dupuis. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Frédéric Dupuis
+-/
+
+import algebra.module.basic
+import ring_theory.ring_invo
+import data.complex.is_R_or_C
+
+/-!
+# Conjugate semimodules
+
+Given a `semimodule R E` with `R` a commutative semiring, we define its conjugate semimodule with
+respect to a ring involution `I : R ≃+* Rᵒᵖ`, in which the scalar product is defined as
+`r • (to_conj I x) = to_conj ((I r) • x)`. This is a generalization of the complex conjugate
+vector space, where the ring involution is the complex conjugation. Since this is an important
+special case, we define also `conj_semimodule 𝕜 E` with `[is_R_or_C 𝕜]` to avoid having to deal
+with the ring involution explicitly.
+
+## Implementation notes
+
+The conjugate semimodule is defined as a copy of the original type, with conversions having to be
+done explicitly via `to_conj` and `from_conj`, as for the opposite type. Facts that are specific
+to normed spaces and inner product spaces are defined in `analysis/normed_space/conjugate.lean`.
+
+## References
+
+* https://en.wikipedia.org/wiki/Complex_conjugate_vector_space
+
+## Tags
+
+conjugate semimodule, conjugate vector space
+-/
+
+variables {R : Type*} [comm_semiring R]
+
+/-- The conjugate of a semimodule `E` with respect to a ring involution `I`. -/
+@[derive [add_comm_monoid], nolint unused_arguments]
+def conjugate_semimodule (I : R ≃+* Rᵒᵖ) (E : Type*) [add_comm_monoid E] := E
+
+instance {I : R ≃+* Rᵒᵖ} {E : Type*} [add_comm_monoid E] : inhabited (conjugate_semimodule I E) :=
+  ⟨0⟩
+
+/-- The `is_R_or_C` complex conjugate semimodule of `E` -/
+abbreviation conj_semimodule (𝕜 : Type*) [is_R_or_C 𝕜] (E : Type*) [add_comm_monoid E] :=
+  conjugate_semimodule (is_R_or_C.conj_to_ring_equiv 𝕜) E
+
+namespace conjugate_semimodule
+
+variables {E : Type*}  [add_comm_monoid E]
+
+/-- The equivalence between `E` and its conjugate semimodule. -/
+@[nolint unused_arguments] def conj_equiv (I : R ≃+* Rᵒᵖ) : E ≃+ conjugate_semimodule I E :=
+{ to_fun := id,
+  inv_fun := id,
+  left_inv := λ x, rfl,
+  right_inv := λ x, rfl,
+  map_add' := λ x y, rfl }
+
+variables {I : R ≃+* Rᵒᵖ}
+
+instance [has_scalar R E] : has_scalar R (conjugate_semimodule I E) :=
+{ smul := λ r x, conj_equiv I ((I r).unop • ((conj_equiv I).symm  x)) }
+
+instance [mul_action R E] : mul_action R (conjugate_semimodule I E) :=
+{ one_smul := λ x, by { change conj_equiv I ((I 1).unop • ((conj_equiv I).symm x)) = x, simp },
+  mul_smul := λ x y z,
+  begin
+    change conj_equiv I( (I (x * y)).unop • ((conj_equiv I).symm z))
+      = conj_equiv I
+       ((I x).unop • ((conj_equiv I).symm ( conj_equiv I ((I y).unop • ((conj_equiv I).symm z)) ))),
+    simp [mul_comm, mul_smul],
+  end }
```

Comment:
I'm a bit puzzled as to why, but this seems to work:
```suggestion
instance [mul_action R E] : mul_action R (conjugate_semimodule I E) :=
{ one_smul := @one_smul R E _ _,
  mul_smul := @mul_smul R E _ _ }
```
and likewise for the others.  (I'm confused because, shouldn't it be necessary to use `mul_comm` at some point, like you do?)

Reply:
OK, there's something *really* weird going on here. If I only make the change for `mul_smul`, it breaks as expected. But if I first change `one_smul`, it accepts the new proof, but then `mul_smul` breaks! And if I change both, Lean accepts the two proofs but then the rest of the file breaks. I think if we write `one_smul := @one_smul R E _ _`, we are coercing from `conjugate_semimodule I E` to `E` at the wrong place, and this somehow propagates to the rest of the file.

Reply:
I think that `one_smul := @one_smul R E _ _` tricks Lean into using the wrong instance of `has_scalar`.

Reply:
I think the proper fix to avoid this sort of stuff is to declare `conjugate_semimodule` to be irreducible after having defined `conj_equiv`, as Scott suggested [here](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/(anti)-monotone).